Fix #82

jbh opened the pull request:
- [x] Are you fixing a bug?
Bug details can be found in #82.

jbh committed:
Add  to PHPDoc.

jbh committed:
Add more missing parameters to PHPDoc.

jbh committed:
Add even more missing parameters to PHPDoc...

jbh committed:
Add ext-json as composer requirement.

jbh committed:
Change template namespace to be dash delimited.

jbh committed:
Modify the CreateTest to check for dash delimited template namespace.

michalbundyra commented on src/CreateHandler/CreateHandler.php:
```diff
@@ -37,10 +37,11 @@ public function __construct(string $classSkeleton = self::CLASS_SKELETON, string
     }
 
     /**
+     * @param string $class
```
Please remove, it is redundant, doesn't give us any additional information than function signature. This is how we are doing now in ZF.

froschdesign replied:
Please undo all these changes as they are not needed. The method declaration already contains this information!
Additionally the topic of this PR is the fix of #82 rather than adding more DocBlocks or clean up anything.

That makes the review easier. Thanks!

michalbundyra commented on src/CreateHandler/CreateHandler.php:
```diff
@@ -37,10 +37,11 @@ public function __construct(string $classSkeleton = self::CLASS_SKELETON, string
      *     are the substitution strings to match, and the values are the associated
      *     values to substitute.
-     * @throws CreateHandlerException
+     * @return string
```
The same about `@return` tag. We have defined RTH with the method declaration so it's useless.

michalbundyra commented on src/CreateHandler/CreateHandler.php:
```diff
@@ -156,6 +159,7 @@ private function discoverNamespaceAndPath(string $class, array $autoloaders) : a
     }
 
     /**
+     * @param string $class
```
Basically revert all changes in that file and everywhere where you changed PHPDocs, please.

michalbundyra commented on src/Module/Create.php:
```diff
@@ -134,8 +134,9 @@ private function createConfigProvider(string $modulePath, string $moduleName) :
-        $namespace = str_replace('\\', '-', $moduleName);
-        $namespace = strtolower($namespace);
-        return $namespace;
+        $pattern     = ['#(?<=(?:\p{Lu}))(\p{Lu}\p{Ll})#', '#(?<=(?:\p{Ll}|\p{Nd}))(\p{Lu})#'];
```
I think these regular expressions are not very clear. Maybe will be worth to add some comment here? Why we need two? Is it the simplest way to do what we need? Not sure, but maybe will be better to use `CamelCaseToDash` filter from `zend-filter`?

jbh replied:
If we comment this and use `zend-filter` instead, we'll need to refactor the `TemplateResolutionTrait.php` also.

michalbundyra replied:
Ok, I see, it is the same as there, so maybe we can reuse it somehow?

jbh replied:
I had wanted to, but I couldn't think of a graceful way of doing it. `TemplateResolutionTrait.php` was meant for `CreateHandler`. I suppose I could still use it in `Module`, but it just felt wrong to do so.

jbh replied:
Is it not a faux pas to use `CreateHandler\TemplateResolutionTrait` in `Module\Create`? If it isn't, I'll do that instead.

michalbundyra replied:
Ok, let leave it as it is, and we will see what @weierophinney is gonna say. I would like to re-use the same functionality. I can see couple options, and the easiest would be extracting this function to separate Trait and reuse it in both places.

jbh replied:
Yes, those were my first thoughts as well, but I was trying to get this in working order as soon as possible, cause it's broken right now. :)

michalbundyra replied:
Sure, leave it for now, and we will see what @weierophinney decide.

jbh replied:
And if we separate it to another trait, where will that live? I suppose we'll need a `Trait` directory for common traits. I'll wait for @weierophinney to weigh in.

weierophinney replied:
I'd put common traits in the source root, honestly, since each command is in its own subdirectory.

michalbundyra commented on test/Module/CreateTest.php:
```diff
@@ -129,7 +129,7 @@ public function testCreatesConfigProvider()
         $this->assertSame(1, preg_match('/\bclass ConfigProvider\b/', $configProviderContent));
         $command = $this->command;
-        $expectedContent = sprintf($command::TEMPLATE_CONFIG_PROVIDER, 'MyApp', 'myapp');
+        $expectedContent = sprintf($command::TEMPLATE_CONFIG_PROVIDER, 'MyApp', 'my-app');
```
I am not sure if that test case covers all changes in `createTemplateNamespace`. Maybe we need to add something more (we have there two regular expressions?).

jbh replied:
This test is just testing if the ConfigProvider is the right format. There could be more tests for template namespace.

weierophinney replied:
I'd say that once you extract the trait, you should add the additional tests as well.

jbh committed:
Remove unnecessary @param PHPDocs.

michalbundyra commented on src/CreateHandler/CreateHandler.php:
```diff
@@ -71,9 +70,6 @@ public function process(
     }
 
-    /**
-     * @throws CreateHandlerException
```
Revert this change - see line 96.

jbh committed:
Add back @throws.

weierophinney reviewed (changes requested):
Extract the trait to the source root, and add the tests as indicated by @webimpress , and I think this is good to go.

Thanks, @jbh !

jbh commented:
Will do. :)

jbh committed:
Move TemplateResolutionTrait to src root.

jbh committed:
Use TemplateResolutionTrait in Module.

jbh commented:
I'll write tests either tonight or in the morning, but I wanted feedback on moving the entire trait to the source root. I just couldn't find a good reason to create a new trait and move just the one method alone.

jbh committed:
Add new tests for module template path name.

jbh commented:
So the regex as I understand it works as such. The first, `(?<=(?:\p{Lu}))(\p{Lu}\p{Ll})`, matches indefinite uppercase and uppercase-lowercase. It will match `LLL...L` and `Ll`. The second, `(?<=(?:\p{Ll}|\p{Nd}))(\p{Lu})`, is for module names with numbers in them. I believe the two test cases I added will cover these. I might have missed a test, but I can't think of anymore currently.

weierophinney committed:
qa: Import all traits, update copyrights

Imports the TemplateResolutionTrait in classes that need it (instead of
using global resolution).

Updates copyright range in all changed files.

weierophinney commented:
@jbh — I've reset this against develop, as it's a slight behavior change, and it's best to call those out at minor releases. I rebased your branch, and then applied one fix (we import all classes, interfaces, and traits used normally).

Thanks a ton for this! I also got to learn about unicode `\p{}` characters sets in the process, which was awesome!
